altair/vegalite/v5/api.py

```diff
@@ -3865,101 +3850,81 @@ def interactive(  # noqa: C901
+                    scale = {"domain": nonreactive_domain}
                 else:
-                    legend_encoding_type = interactive_chart.encoding[legend_encoding][
-                        "type"
-                    ]
-                if (
-                    legend_encoding_type == "nominal"
-                ):  # TODO Ideally this would work for ordinal data too
-                    legend_selection = selection_point(
-                        bind="legend", encodings=[legend_encoding]
-                    )
-                    initial_computed_domain = param(expr=f"domain('{legend_encoding}')")
-                    nonreactive_domain = param(
-                        react=False, expr=initial_computed_domain.name
-                    )
-                    if utils.is_undefined(
-                        interactive_chart.encoding[legend_encoding]["scale"]
-                    ):
-                        interactive_chart.encoding[legend_encoding]["scale"] = {
-                            "domain": nonreactive_domain
-                        }
-                    else:
-                        interactive_chart.encoding[legend_encoding]["scale"][
-                            "domain"
-                        ] = nonreactive_domain
-
-                    interactive_chart = interactive_chart.add_params(
-                        legend_selection, initial_computed_domain, nonreactive_domain
-                    ).transform_filter(legend_selection)
-        return interactive_chart
+                    scale["domain"] = nonreactive_domain
+                chart = chart.add_params(
+                    legend_selection,
+                    initial_computed_domain,
+                    nonreactive_domain,
+                ).transform_filter(legend_selection)
+            else:
+                msg = f"Expected only 'nominal' legend type but got {legend_type!r}"
+                raise NotImplementedError(msg)
+        return chart
+
+
+LegendChannel_T: TypeAlias = Literal[
+    "color",
+    "fill",
+    "shape",
+    "stroke",
+    "opacity",
+    "fillOpacity",
+    "strokeOpacity",
+    "strokeWidth",
+    "strokeDash",
+    "angle",  # TODO Untested
+    "radius",  # TODO Untested
+    "radius2",  # TODO Untested
+    # "size",  # TODO Currently size is not working, renders empty legend
+]
+
+
+def _add_tooltip(chart: _TChart, /) -> _TChart:
+    mark = chart.mark
+    if isinstance(mark, str):
+        mark = {"type": mark, "tooltip": True}
+    else:
+        mark.tooltip = True
+    return chart
+
+
+def _infer_legend_encoding(encoding: FacetedEncoding, /) -> LegendChannel_T:
+    """Set the legend to commonly used encodings by default."""
+    _channels = t.get_args(LegendChannel_T)
+    it = (ch for ch in _channels if not utils.is_undefined(encoding[ch]))
+    if legend := next(it, None):
+        return legend
+    else:
+        msg = f"Unable to infer target channel for 'legend'.\n\n{encoding!r}"
```

Comment:
It wasn't clear to me what the intention was for this case. 

At a minimum I think there should be a warning displayed, rather than silently continuing.

Reply:
I'm not sure what is the best approach here. I agree that a warning/error is more explicit than silently continuing, but since `.interactive()` is a sort of convenient method to quickly add basic interactivity to charts, I also want to keep it simple to use. To me, the least friction for the user is if legend interactivity is applied when it makes sense and not applied when it doesn't make sense without being too noisy with warnings and errors. I'm thinking that people would not expect a gradient color legend to be interactive in any way, but they would expect to click on discrete points/lines to select and unselect groups. This is also how e.g. plotly works, see the examples on this page https://plotly.com/python/plotly-express/#gallery. If we go this path, one case that might be confusing in altair is [quantitative scales in the size encoding](https://altair-viz.github.io/gallery/histogram_scatterplot.html) since these appear to have discrete categories in the legend but clicking them would not select anything (since there are rarely observations with those exact values). I think this is minor, and am still leaning towards the "apply interactivity where it makes sense" approach for `.interactive()` and I'm ok with this method being less explicitly verbose since I see it as more of a convenience.

Reply:
>  If I'm a new user and I call Chart.interactive(legend=True), but result doesn't contain a legend, how am I supposed to know what went wrong?

Good point, and it brings up an interesting distinction I didn't think of before. Whereas setting `tooltip=True` actually **creates** a tooltip, setting `legend=True` converts an existing legend into an interactive/clickable format (but doesn't create one from scratch as you pointed out). I will make this distinction explicit via the docstring and/or a more suitable parameter name. Then we can consider whether that is clear enough or we also need the warning/error; I appreciate your reasoning and suggestion with the nudging message, but I want to avoid that users need to pass `legend=False` just because there color scale is using a quantitative encoding (so eventually I will likely favor either a warning or an explicit docstring, but I will leave the error raise for now until next time I can work on this PR)

Reply:
That's fair!

> > If I'm a new user and I call Chart.interactive(legend=True), but **result** doesn't contain a legend, how am I supposed to know what went wrong?

> Good point, and it brings up an interesting distinction I didn't think of before. Whereas setting `tooltip=True` actually **creates** a tooltip, setting `legend=True` converts an existing legend into an interactive/clickable format (but doesn't create one from scratch as you pointed out)

My bad for writing the last comment without testing.
I was assuming it created a legend (forgetting that it would already be there).

I guess the starting point would be the same though, encountering that error currently would require `legend=True` and a `Chart` without a channel displaying a legend?